src/colmap/estimators/bundle_adjustment.cc

```diff
@@ -810,6 +813,257 @@ void RigBundleAdjuster::ParameterizeCameraRigs(Reconstruction* reconstruction) {
+  ParameterizePoints(reconstruction);
+}
+
+void PositionPriorBundleAdjuster::AddImageToProblem(
+    image_t image_id,
+    Reconstruction* reconstruction,
+    ceres::LossFunction* loss_function,
+    ceres::LossFunction* prior_loss_function) {
+  Image& image = reconstruction->Image(image_id);
+  Camera& camera = reconstruction->Camera(image.CameraId());
+
+  // CostFunction assumes unit quaternions.
+  image.CamFromWorld().rotation.normalize();
+
+  double* cam_from_world_rotation =
+      image.CamFromWorld().rotation.coeffs().data();
+  double* cam_from_world_translation = image.CamFromWorld().translation.data();
+  double* camera_params = camera.params.data();
+
+  const bool constant_cam_pose =
+      !options_.refine_extrinsics || config_.HasConstantCamPose(image_id);
+
+  // Add residuals to bundle adjustment problem.
+  size_t num_observations = 0;
+  for (const Point2D& point2D : image.Points2D()) {
+    if (!point2D.HasPoint3D()) {
+      continue;
+    }
+
+    num_observations += 1;
+    point3D_num_observations_[point2D.point3D_id] += 1;
+
+    Point3D& point3D = reconstruction->Point3D(point2D.point3D_id);
+    assert(point3D.track.Length() > 1);
+
+    if (constant_cam_pose) {
+      problem_->AddResidualBlock(
+          CameraCostFunction<ReprojErrorConstantPoseCostFunction>(
+              camera.model_id, image.CamFromWorld(), point2D.xy),
+          loss_function,
+          point3D.xyz.data(),
+          camera_params);
+    } else {
+      problem_->AddResidualBlock(CameraCostFunction<ReprojErrorCostFunction>(
+                                     camera.model_id, point2D.xy),
+                                 loss_function,
+                                 cam_from_world_rotation,
+                                 cam_from_world_translation,
+                                 point3D.xyz.data(),
+                                 camera_params);
+    }
+  }
+
+  if (num_observations > 0) {
+    camera_ids_.insert(image.CameraId());
+
+    // Set pose parameterization.
+    if (!constant_cam_pose) {
+      SetQuaternionManifold(problem_.get(), cam_from_world_rotation);
+      if (config_.HasConstantCamPositions(image_id)) {
+        const std::vector<int>& constant_position_idxs =
+            config_.ConstantCamPositions(image_id);
+        SetSubsetManifold(3,
+                          constant_position_idxs,
+                          problem_.get(),
+                          cam_from_world_translation);
+      }
+    }
+
+    if (prior_options_.use_prior_position && image.HasPosePrior()) {
+      problem_->AddResidualBlock(PositionPriorErrorCostFunction::Create(
+                                     image.WorldFromCamPrior().position,
+                                     prior_options_.prior_position_covariance),
+                                 prior_loss_function,
+                                 cam_from_world_rotation,
+                                 cam_from_world_translation);
+    }
+  }
```

Comment:
Only this section is different from `BundleAdjuster::AddImageToProblem`, could we reuse the code of this this parent function by calling it first?

Reply:
Done.  We are now calling a AddPosePriorToProblem function after a call to setUpProblem.